Clone the vals map for every path to avoid mutation

waveywaves opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/master/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
This PR fixes https://github.com/helm/helm/issues/6079 
`vals` was getting mutated when multiple charts were provided for linting.
Cloning `vals` for every `path` sees to it that every `path` has it's own unmutated set of `vals`.

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

waveywaves committed:
Clone the vals map for every path to avoid mutation

Signed-off-by: Vibhav Bobade <vibhav.bobde@gmail.com>

waveywaves commented:
cc @karuppiah7890 @adamreese @mattfarina

karuppiah7890 commented:
@waveywaves could you add a failing test for the bug and then pass it with the current code? Tests would give confidence on the robustness of the features

waveywaves committed:
Test for Linting multiple charts with the same vals instance

Signed-off-by: Vibhav Bobade <vibhav.bobde@gmail.com>

waveywaves commented:
@karuppiah7890 Have updated the PR. Do take a look and let me know if I need to make any changes.

karuppiah7890 commented:
@waveywaves I ran the test without the fix and it passed, may be use existing chart / add some chart that when linted after another one breaks the linting due to the issue? 

For example, in first chart, use a chart with a `config` value which is yaml and has `toYaml` used in the templates to use the `config`, in second chart, use `config` value which is a string and used directly with `.Values.config` with no `toYaml` or anything. The example is a subset of the of the charts mentioned in the issue which caused the bug.

karuppiah7890 commented:
Also, do sign off the commit https://github.com/helm/helm/pull/6241/checks?check_run_id=195950295 for the Developer Certificate of Origin check

waveywaves commented:
@karuppiah7890 I have added the testcharts against which I was able to have the test fail before the changes. And I have updated the test as well. Let me know if I need to make any changes to this.

waveywaves committed:
Fix Adding Errors from Linter.Messages to result.Errors

This also fixes <nil> beig added to result.Errors

Signed-off-by: Vibhav Bobade <vibhav.bobde@gmail.com>

karuppiah7890 commented:
I just glanced it. It's awesome 😁. I see you have fixed a Fixme comment too, I have no idea about that, will have to read and understand that. Other than that LGTM. And I can't merge PRs, so cc'ing people. cc @bacongobbler

thomastaylor312 commented on pkg/action/lint.go:
```diff
@@ -59,15 +59,22 @@ func (l *Lint) Run(paths []string, vals map[string]interface{}) *LintResult {
 			if err == errLintNoChart {
 				result.Errors = append(result.Errors, err)
 			}
+			if linter.HighestSeverity >= lowestTolerance {
```
These changes feel out of scope for this PR because it doesn't seem to be related to the values map mutation. Please let me know if I am wrong

waveywaves replied:
None of the `Messages` which are errors are being appended to `Result.Errors` which shouldn't be the case.

waveywaves replied:
While writing the test, I noticed that there was <nil> getting appended to Result.Errors. This was creating an issue because, Result.Errors was supposed tohave a length of zero, instead it had a length of 1, hence failing my test even on the right code.

When running lintChart on a chart which passes the lint, it would return <nil> for err which is expected. But this lint was being appended to result.Errors which isn't the expected behaviour. Do correct me if I'm wrong.

thomastaylor312 replied:
OK, that make sense then. So this was a different bug you found along the way and fixed also

thomastaylor312 replied:
@waveywaves could you write a unit test to make sure we can validate that in the future?

waveywaves replied:
@thomastaylor312 I have updated the PR, can you check and let me know.

thomastaylor312 reviewed (commented):
This is looking good. Thanks for adding a good unit test for this! Just one comment around whether a block of changes is needed for this PR

thomastaylor312 reviewed (approved):
Thanks for the great work on this!
